pkg/controller/clusterpool/collections.go

```diff
@@ -675,94 +692,281 @@ func getAllCustomizationsForPool(c client.Client, pool *hivev1.ClusterPool, logg
-					return err
-				}
-			}
+	for _, cd := range cds.byCDName {
+		// CD has CDC
+		cpRef := cd.Spec.ClusterPoolRef
+		if cpRef.CustomizationRef == nil {
+			continue
 		}
-	}
 
-	// Notice a Broken CD => update the CDC's LastApplyStatus to BrokenByCloud;
-	for _, cd := range cds.broken {
-		cdcRef := cd.Spec.ClusterPoolRef.CustomizationRef
-		if cdcRef == nil {
+		// CDC exists
+		cdc, ok := cdcs.namespace[cpRef.CustomizationRef.Name]
+		if !ok {
 			continue
 		}
 
-		update := true
-		for _, cdc := range cdcs.cloud {
-			if cdc.Name == cdcRef.Name {
-				update = false
-				break
+		// CDC is no reserved
+		available := conditionsv1.FindStatusCondition(cdc.Status.Conditions, conditionsv1.ConditionAvailable)
+		if available == nil || available.Status == corev1.ConditionTrue {
+			// Fix CDC availability
+			cdc.Status.ClusterDeploymentRef = &corev1.LocalObjectReference{Name: cd.Name}
+			cdc.Status.ClusterPoolRef = &corev1.LocalObjectReference{Name: pool.Name}
+			if err := cdcs.Reserve(c, cdc); err != nil {
+				return err
 			}
 		}
+	}
 
-		if update {
-			cdc := cdcs.byCDCName[cd.Spec.ClusterPoolRef.CustomizationRef.Name]
-			cdcs.cloud = append(cdcs.cloud, cdc)
-			cdc.Status.LastApplyStatus = hivev1.LastApplyBrokenCloud
-			cdc.Status.LastApplyTime = metav1.Now()
-			if err := c.Status().Update(context.TODO(), cdc); err != nil {
+	// Notice a Broken CD => update the CDC's ApplySucceded condition to BrokenByCloud;
+	for _, cd := range cds.Broken() {
+		cdcRef := cd.Spec.ClusterPoolRef.CustomizationRef
+		if cdcRef == nil {
+			continue
+		}
+		if cdc := cdcs.byCDCName[cdcRef.Name]; cdc != nil {
+			if err := cdcs.BrokenCloud(c, cdc); err != nil {
 				return err
 			}
 		}
 	}
-	// Notice a CD has finished installing => update the CDC's LastApplyStatus to Success;
-	for _, cd := range cds.Unassigned(false) {
+
+	// Notice a CD has finished installing => update the CDC's ApplySucceeded condition to Success;
+	for _, cd := range cds.Installed() {
 		if cd.Spec.ClusterPoolRef.CustomizationRef == nil {
 			continue
 		}
 		if cdc, ok := cdcs.byCDCName[cd.Spec.ClusterPoolRef.CustomizationRef.Name]; ok {
-			if cdc.Status.LastApplyStatus != hivev1.LastApplySucceeded {
-				cdc.Status.LastApplyStatus = hivev1.LastApplySucceeded
-				cdc.Status.LastApplyTime = metav1.Now()
-				if err := c.Status().Update(context.TODO(), cdc); err != nil {
-					return err
-				}
+			if err := cdcs.Succeeded(c, cdc); err != nil {
+				return err
 			}
 		}
 	}
 
-	// Update Cluster Pool Inventory Condition
+	if err := cdcs.UpdateInventoryValidCondition(c, pool); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Update Cluster Pool Inventory Condition
+func (cdcs *cdcCollection) UpdateInventoryValidCondition(c client.Client, pool *hivev1.ClusterPool) error {
 	message := ""
 	status := corev1.ConditionTrue
 	reason := hivev1.InventoryReasonValid
 	if (len(cdcs.syntax) + len(cdcs.cloud) + len(cdcs.missing)) > 0 {
+		var _ json.Marshaler = &cdcCollection{}
```

Comment:
This should be at the top level of the file, somewhere near the `Marshal` func that makes it work.